
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
      
        <link rel="prev" href="..">
      
      
        <link rel="next" href="../support/">
      
      <link rel="icon" href="../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.4.2, mkdocs-material-9.1.0">
    
    
      
        <title>Contribute - TB-ML-Containers</title>
      
    
    
      <link rel="stylesheet" href="../assets/stylesheets/main.402914a4.min.css">
      
        
        <link rel="stylesheet" href="../assets/stylesheets/palette.a0c5b2b5.min.css">
      
      

    
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
    <script>__md_scope=new URL("..",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="" data-md-color-accent="">
  
    
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#contribute" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href=".." title="TB-ML-Containers" class="md-header__button md-logo" aria-label="TB-ML-Containers" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54Z"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            TB-ML-Containers
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Contribute
            
          </span>
        </div>
      </div>
    </div>
    
    
    
      <label class="md-header__button md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href=".." title="TB-ML-Containers" class="md-nav__button md-logo" aria-label="TB-ML-Containers" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54Z"/></svg>

    </a>
    TB-ML-Containers
  </label>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href=".." class="md-nav__link">
        TB-ML-Containers
      </a>
    </li>
  

    
      
      
      

  
  
    
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          Contribute
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        Contribute
      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#how-to-develop-a-container" class="md-nav__link">
    How to develop a container
  </a>
  
    <nav class="md-nav" aria-label="How to develop a container">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#step-1-training-the-model" class="md-nav__link">
    Step 1 - Training the model
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#step-2-set-up-your-prediction-script" class="md-nav__link">
    Step 2 - Set up your prediction script
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#step-3-create-container" class="md-nav__link">
    Step 3 - Create container
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#adding-your-container-to-tb-ml-containers" class="md-nav__link">
    Adding your container to TB-ML-Containers
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../support/" class="md-nav__link">
        Support
      </a>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
      
      
      <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4" >
      
      
      
        <label class="md-nav__link" for="__nav_4" id="__nav_4_label" tabindex="0">
          Containers
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_4_label" aria-expanded="false">
        <label class="md-nav__title" for="__nav_4">
          <span class="md-nav__icon md-icon"></span>
          Containers
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
      
      
      <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_1" >
      
      
      
        <label class="md-nav__link" for="__nav_4_1" id="__nav_4_1_label" tabindex="0">
          Prediction
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_4_1_label" aria-expanded="false">
        <label class="md-nav__title" for="__nav_4_1">
          <span class="md-nav__icon md-icon"></span>
          Prediction
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../containers/prediction/RF-Streptomycin/" class="md-nav__link">
        RF-Streptomycin
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../containers/prediction/TB-AMR-CNN/" class="md-nav__link">
        TB-AMR-CNN
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../containers/prediction/TB-DR-pred-nn/" class="md-nav__link">
        tb-dr-pred-nn
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../containers/prediction/aggreen-MTB-CNN/" class="md-nav__link">
        aggreen-MTB-CNN
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../containers/prediction/mykrobe/" class="md-nav__link">
        mykrobe
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

            
          
            
              
  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
      
      
      <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_2" >
      
      
      
        <label class="md-nav__link" for="__nav_4_2" id="__nav_4_2_label" tabindex="0">
          Preprocessing
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_4_2_label" aria-expanded="false">
        <label class="md-nav__title" for="__nav_4_2">
          <span class="md-nav__icon md-icon"></span>
          Preprocessing
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../containers/preprocessing/aln2onehot/" class="md-nav__link">
        aln2one-hot
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../containers/preprocessing/aln2variants/" class="md-nav__link">
        aln2variants
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../containers/preprocessing/raw2consensus/" class="md-nav__link">
        raw2consensus
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../containers/preprocessing/raw2onehot/" class="md-nav__link">
        raw2one-hot
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#how-to-develop-a-container" class="md-nav__link">
    How to develop a container
  </a>
  
    <nav class="md-nav" aria-label="How to develop a container">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#step-1-training-the-model" class="md-nav__link">
    Step 1 - Training the model
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#step-2-set-up-your-prediction-script" class="md-nav__link">
    Step 2 - Set up your prediction script
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#step-3-create-container" class="md-nav__link">
    Step 3 - Create container
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#adding-your-container-to-tb-ml-containers" class="md-nav__link">
    Adding your container to TB-ML-Containers
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  


<h1 id="contribute">Contribute</h1>
<h2 id="how-to-develop-a-container">How to develop a container</h2>
<p>A container is essentially a mini virtual machine that contains all software needed to run predictions on new data.
<a href="https://www.docker.com/">Docker</a> is used to build the containers and they can then be run on any system which has docker installed. There are a few steps required to build a container. We'll go through these by going through a whole example, from training to prediction. </p>
<h3 id="step-1-training-the-model">Step 1 - Training the model</h3>
<p>As an example we will train a random forest classifier to predict resistance to rifampicin. The input data is a set of 100 resistant and 100 sensitive isolates randomly taken from the SRA. The input data and scripts are available from <a href="https://github.com/jodyphelan/simple-rif-rf">this repository</a>. </p>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>This is a very quick and dirty model that takes a lot of shortcuts, it is just meant for demonstration purposes.</p>
</div>
<p>First lets set up a conda environment for our code:</p>
<div class="highlight"><pre><span></span><code>mamba create -n ml -c conda-forge -c bioconda scikit-learn bcftools
conda activate ml
</code></pre></div>
<p>This is our training code.</p>
<div class="highlight"><pre><span></span><code><span class="kn">from</span> <span class="nn">sklearn.ensemble</span> <span class="kn">import</span> <span class="n">RandomForestClassifier</span>
<span class="kn">import</span> <span class="nn">pickle</span>
<span class="kn">import</span> <span class="nn">subprocess</span> <span class="k">as</span> <span class="nn">sp</span>
<span class="kn">from</span> <span class="nn">collections</span> <span class="kn">import</span> <span class="n">defaultdict</span>

<span class="n">genos</span> <span class="o">=</span> <span class="n">defaultdict</span><span class="p">(</span><span class="nb">dict</span><span class="p">)</span>

<span class="c1"># get genotypes for all samples at all positions in the vcf</span>
<span class="k">for</span> <span class="n">l</span> <span class="ow">in</span> <span class="n">sp</span><span class="o">.</span><span class="n">Popen</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;bcftools query -f &#39;[%POS\t%SAMPLE\t</span><span class="si">%G</span><span class="s2">T\n]&#39; snps.vcf.gz&quot;</span><span class="p">,</span><span class="n">shell</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span><span class="n">stdout</span><span class="o">=</span><span class="n">sp</span><span class="o">.</span><span class="n">PIPE</span><span class="p">)</span><span class="o">.</span><span class="n">stdout</span><span class="p">:</span>
    <span class="c1"># l is a byte string, so we need to decode it to a string and strip the newline character at the end</span>
    <span class="n">pos</span><span class="p">,</span><span class="n">sample</span><span class="p">,</span><span class="n">gt</span> <span class="o">=</span> <span class="n">l</span><span class="o">.</span><span class="n">decode</span><span class="p">()</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span><span class="o">.</span><span class="n">split</span><span class="p">()</span>
    <span class="c1"># convert the position to an integer</span>
    <span class="n">pos</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">pos</span><span class="p">)</span>
    <span class="c1"># if the genotype is missing or reference, set it to 0 (otherwise set it to 1)</span>
    <span class="k">if</span> <span class="n">gt</span><span class="o">==</span><span class="s2">&quot;./.&quot;</span> <span class="ow">or</span> <span class="n">gt</span><span class="o">==</span><span class="s2">&quot;0/0&quot;</span><span class="p">:</span>
        <span class="n">gt</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">gt</span> <span class="o">=</span> <span class="mi">1</span>
    <span class="c1"># add the genotype to the dictionary</span>
    <span class="n">genos</span><span class="p">[</span><span class="n">sample</span><span class="p">][</span><span class="n">pos</span><span class="p">]</span> <span class="o">=</span> <span class="n">gt</span>

<span class="c1"># get the phenotype for each sample</span>
<span class="n">pheno</span> <span class="o">=</span> <span class="p">{}</span>
<span class="k">for</span> <span class="n">l</span> <span class="ow">in</span> <span class="nb">open</span><span class="p">(</span><span class="s2">&quot;pheno.txt&quot;</span><span class="p">):</span>
    <span class="n">row</span> <span class="o">=</span> <span class="n">l</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span><span class="o">.</span><span class="n">split</span><span class="p">()</span>
    <span class="n">pheno</span><span class="p">[</span><span class="n">row</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>

<span class="c1"># convert the genotype dictionary to a list of lists</span>
<span class="n">X</span> <span class="o">=</span> <span class="p">[</span><span class="nb">list</span><span class="p">(</span><span class="n">genos</span><span class="p">[</span><span class="n">s</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">())</span> <span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="n">pheno</span><span class="p">]</span> 
<span class="n">y</span> <span class="o">=</span> <span class="p">[</span><span class="n">pheno</span><span class="p">[</span><span class="n">s</span><span class="p">]</span> <span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="n">pheno</span><span class="p">]</span>
<span class="c1"># use scikit-learn to train a random forest classifier</span>
<span class="n">clf</span> <span class="o">=</span> <span class="n">RandomForestClassifier</span><span class="p">(</span><span class="n">max_depth</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">random_state</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="n">clf</span><span class="o">.</span><span class="n">fit</span><span class="p">(</span><span class="n">X</span><span class="p">,</span> <span class="n">y</span><span class="p">)</span>

<span class="c1"># dump the model and the positions to a pickle file</span>
<span class="n">positions</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">genos</span><span class="o">.</span><span class="n">values</span><span class="p">())[</span><span class="mi">0</span><span class="p">])</span>
<span class="n">pickle</span><span class="o">.</span><span class="n">dump</span><span class="p">({</span><span class="s2">&quot;model&quot;</span><span class="p">:</span><span class="n">clf</span><span class="p">,</span><span class="s2">&quot;positions&quot;</span><span class="p">:</span><span class="n">positions</span><span class="p">},</span> <span class="nb">open</span><span class="p">(</span><span class="s2">&quot;model.pkl&quot;</span><span class="p">,</span><span class="s2">&quot;wb&quot;</span><span class="p">))</span>
</code></pre></div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>We have used pickle to dump the model to a file. Importantly, we also include the genomic positions of all the features so it will allow us to re-create the input data in the exact same format. We can then use the <code>.predict()</code> method from the object when we reload it somewhere else.</p>
</div>
<p>Execute the code to run the training:</p>
<div class="highlight"><pre><span></span><code>$<span class="w"> </span>python<span class="w"> </span>train.py
</code></pre></div>
<h3 id="step-2-set-up-your-prediction-script">Step 2 - Set up your prediction script</h3>
<p>Now that we have the model saved, we can create a script to load it along with a vcf from a new sample and make a prediction. Here is the code for that.</p>
<div class="highlight"><pre><span></span><code><span class="sd">&quot;&quot;&quot; read in genotypes from VCF file and predict outcome &quot;&quot;&quot;</span>
<span class="kn">import</span> <span class="nn">sys</span>
<span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">argparse</span>
<span class="kn">import</span> <span class="nn">pickle</span>
<span class="kn">import</span> <span class="nn">subprocess</span> <span class="k">as</span> <span class="nn">sp</span>

<span class="c1"># parse command line arguments</span>
<span class="n">parser</span> <span class="o">=</span> <span class="n">argparse</span><span class="o">.</span><span class="n">ArgumentParser</span><span class="p">(</span><span class="n">description</span><span class="o">=</span><span class="s1">&#39;Predict outcome from VCF file&#39;</span><span class="p">)</span>
<span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s1">&#39;--vcf&#39;</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s1">&#39;VCF file&#39;</span><span class="p">,</span> <span class="n">required</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s1">&#39;--model&#39;</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s1">&#39;model file&#39;</span><span class="p">,</span> <span class="n">required</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">args</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">parse_args</span><span class="p">()</span>

<span class="c1"># load model</span>
<span class="n">items</span> <span class="o">=</span> <span class="n">pickle</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="nb">open</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">model</span><span class="p">,</span> <span class="s2">&quot;rb&quot;</span><span class="p">))</span>
<span class="n">clf</span> <span class="o">=</span> <span class="n">items</span><span class="p">[</span><span class="s1">&#39;model&#39;</span><span class="p">]</span>
<span class="n">positions</span> <span class="o">=</span> <span class="n">items</span><span class="p">[</span><span class="s1">&#39;positions&#39;</span><span class="p">]</span>



<span class="c1"># extract genotypes</span>
<span class="n">genos</span> <span class="o">=</span> <span class="p">{}</span>
<span class="k">for</span> <span class="n">l</span> <span class="ow">in</span> <span class="n">sp</span><span class="o">.</span><span class="n">Popen</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;bcftools query -f &#39;%POS[\t</span><span class="si">%G</span><span class="s2">T\n]&#39; &quot;</span><span class="o">+</span> <span class="n">args</span><span class="o">.</span><span class="n">vcf</span><span class="p">,</span><span class="n">shell</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span><span class="n">stdout</span><span class="o">=</span><span class="n">sp</span><span class="o">.</span><span class="n">PIPE</span><span class="p">)</span><span class="o">.</span><span class="n">stdout</span><span class="p">:</span>
    <span class="n">row</span> <span class="o">=</span> <span class="n">l</span><span class="o">.</span><span class="n">decode</span><span class="p">()</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span><span class="o">.</span><span class="n">split</span><span class="p">()</span>
    <span class="k">if</span> <span class="n">row</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">!=</span><span class="s2">&quot;0/0&quot;</span><span class="p">:</span>
        <span class="n">genos</span><span class="p">[</span><span class="nb">int</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="mi">0</span><span class="p">])]</span> <span class="o">=</span> <span class="mi">1</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">genos</span><span class="p">[</span><span class="nb">int</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="mi">0</span><span class="p">])]</span> <span class="o">=</span> <span class="mi">0</span>

<span class="c1"># fill in missing genotypes with 0 (ref)</span>
<span class="k">for</span> <span class="n">p</span> <span class="ow">in</span> <span class="n">positions</span><span class="p">:</span>
    <span class="k">if</span> <span class="n">p</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">genos</span><span class="p">:</span>
        <span class="n">genos</span><span class="p">[</span><span class="n">p</span><span class="p">]</span> <span class="o">=</span> <span class="mi">0</span>

<span class="n">genos</span> <span class="o">=</span> <span class="p">[[</span><span class="n">genos</span><span class="p">[</span><span class="n">p</span><span class="p">]</span> <span class="k">for</span> <span class="n">p</span> <span class="ow">in</span> <span class="n">positions</span><span class="p">]]</span>

<span class="c1"># predict outcome</span>
<span class="n">pred</span> <span class="o">=</span> <span class="n">clf</span><span class="o">.</span><span class="n">predict</span><span class="p">(</span><span class="n">genos</span><span class="p">)</span>
<span class="c1"># write output</span>
<span class="n">sys</span><span class="o">.</span><span class="n">stdout</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;drug,prediction</span><span class="se">\n</span><span class="s2">rifampicin,</span><span class="si">%s</span><span class="se">\n</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="n">pred</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
</code></pre></div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>If you are making your container to be complient with <code>tb-ml</code> you need to write any results you want in the final output to stdout.</p>
</div>
<p>We can then use this script to predict on some new data.</p>
<div class="highlight"><pre><span></span><code>$<span class="w"> </span>python<span class="w"> </span>predict.py<span class="w"> </span>--vcf<span class="w"> </span>por5A1.vcf.gz<span class="w"> </span>--model<span class="w"> </span>model.pkl
drug,prediction
rifampicin,1
</code></pre></div>
<p>Now we are ready to package our prediction pipeline into a container.</p>
<h3 id="step-3-create-container">Step 3 - Create container</h3>
<p>To build a container you first need docker installed. Head over to https://www.docker.com/ to download the latest version. Once you have it installed you can start the build process. First we need to define what software we need.
It is very important to use the exact same versions of libraries and software in the docker container as you used to train. You can check this with conda.</p>
<div class="highlight"><pre><span></span><code>$<span class="w"> </span>conda<span class="w"> </span>list<span class="w"> </span><span class="p">|</span><span class="w"> </span>grep<span class="w"> </span>bcftools
bcftools<span class="w">                  </span><span class="m">1</span>.16<span class="w">                 </span>h83fc8ca_1<span class="w">    </span>bioconda
$<span class="w"> </span>conda<span class="w"> </span>list<span class="w"> </span><span class="p">|</span><span class="w"> </span>grep<span class="w"> </span>scikit-learn
scikit-learn<span class="w">              </span><span class="m">1</span>.2.1<span class="w">           </span>py311h087fafe_0<span class="w">    </span>conda-forge
</code></pre></div>
<p>So we can see we need to use <code>bcftools=1.16</code> and <code>scikit-learn=1.2.1</code>. To build our container we need to create a file called <code>Dockerfile</code> and insert our software dependencies. We also need to add our saved model file and the prediction script. </p>
<div class="highlight"><pre><span></span><code>FROM mambaorg/micromamba:latest
LABEL image.name=&quot;jodyphelan/simple-rif-rf&quot;

# This needs to be set for the PATH variable to be set correctly
ARG MAMBA_DOCKERFILE_ACTIVATE=1

# Install our software
RUN micromamba install -y -n base -c bioconda -c conda-forge \
    scikit-learn=1.2.1 \
    bcftools=1.16 \
    tqdm &amp;&amp; \
    micromamba clean --all --yes

# create a directory for the internal data used by the container
USER root
RUN mkdir /internal_data /data

# copy the model, data files, and scripts
COPY model.pkl /internal_data/model.pkl
COPY predict.py /internal_data

# set `/data` as working directory so that the output is written to the
# mount point when run with `docker run -v $PWD:/data ... -o output.csv`
WORKDIR /data

ENTRYPOINT [&quot;/usr/local/bin/_entrypoint.sh&quot;, &quot;python&quot;, &quot;/internal_data/predict.py&quot;, &quot;--model&quot;, &quot;/internal_data/model.pkl&quot;]
</code></pre></div>
<p>We can then build the container with </p>
<div class="highlight"><pre><span></span><code>$<span class="w"> </span>docker<span class="w"> </span>build<span class="w"> </span>-t<span class="w"> </span>simple-rif-rf<span class="w"> </span>.<span class="w"> </span>
</code></pre></div>
<p>In a nutshell, this installs our required software, copies the data files and scripts to the container and finally defines an "entrypoint" script that will be run when the container is executed.</p>
<p>We can then use this container in a similar way to how we used the script to predict. Since, we already point the <code>--model</code> flag to the right file in the ENTRYPOINT variable we only need to give the <code>--vcf</code> argument.</p>
<div class="highlight"><pre><span></span><code>$<span class="w"> </span>docker<span class="w"> </span>run<span class="w"> </span>--rm<span class="w"> </span>-it<span class="w"> </span>-v<span class="w"> </span><span class="nv">$PWD</span>:/data<span class="w">  </span>simple-rif-rf<span class="w">  </span>
drug,prediction
rifampicin,1
</code></pre></div>
<p>Now finally you can publish your container to dockerhub. Head over to https://hub.docker.com/ and set yourself up with an account. Once you have done this you can link your account with your local docker installation.</p>
<div class="highlight"><pre><span></span><code>$<span class="w"> </span>docker<span class="w"> </span>login
</code></pre></div>
<p>And finally you can push the container to dockerhub</p>
<div class="highlight"><pre><span></span><code>$<span class="w"> </span>docker<span class="w"> </span>push<span class="w"> </span>jodyphelan/simple-rif-rf:latest
</code></pre></div>
<h2 id="adding-your-container-to-tb-ml-containers">Adding your container to TB-ML-Containers</h2>
<p>If you want to publicise your container you can add it to this website to improve its visibility. To do this you just have to fork <a href="https://github.com/TB-ML/tb-ml-containers">this repo</a>. After this, you should add a markdown file to the appropriate location <code>/docs/containers/prediction</code> for prediction containers or <code>/docs/containers/preprocessing</code> for preprocessing containers. The markdown file should contain a few variables that are defined at the top of the page in yaml format. In particular, for prediction containers they need:</p>
<ul>
<li>title - Name of the model</li>
<li>drugs - A list of drugs it predicts for</li>
<li>architecture - The type of ML model</li>
<li>input - The input type it needs (e.g. VCF)</li>
<li>docker - The name of the container on dockerhub</li>
</ul>
<p>After defining these you can write about your model using standard markdown. For example:</p>
<div class="highlight"><pre><span></span><code>---
title: simple-rif-rf
drugs: [&#39;rifampicin&#39;]
architecture: RandomForest
docker: jodyphelan/simple-rif-rf
input: variants/VCF
---

# Rifampicin RandomForest

This container predicts rifampicin resistance from variants in VCF format.
</code></pre></div>
<p>Now just put in a pull request and we will add it in to the website. You model will automatically appear on the tables on the front page. A dedicated rendering your markdown content will also be generated using the markdown filename as the link.</p>





                
              </article>
            </div>
          
          
        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    <script id="__config" type="application/json">{"base": "..", "features": [], "search": "../assets/javascripts/workers/search.208ed371.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script>
    
    
      <script src="../assets/javascripts/bundle.ce72ebac.min.js"></script>
      
    
  </body>
</html>